sound/soc/sof/intel/hda.c

```diff
@@ -66,28 +62,60 @@ int hda_ctrl_dai_widget_setup(struct snd_soc_dapm_widget *w, bool setup)
 
 	/*
 	 * For static pipelines, the DAI widget would already be set up and calling
-	 * sof_widget_setup()/free() simply returns without doing anything.
-	 * For dynamic pipelines, the DAI widget will be set up or freed here.
+	 * sof_widget_setup() simply returns without doing anything.
+	 * For dynamic pipelines, the DAI widget will be set up now.
 	 */
-	if (setup) {
-		ret = sof_widget_setup(sdev, swidget);
-		if (ret < 0) {
-			dev_err(sdev->dev, "error: setting up DAI widget %s\n", w->name);
-			return ret;
-		}
+	ret = sof_widget_setup(sdev, swidget);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed setting up DAI widget %s\n", w->name);
+		return ret;
+	}
+
+	/* send DAI_CONFIG IPC */
+	ret = sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
+				  &reply, sizeof(reply));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed setting DAI config for %s\n", w->name);
+		return ret;
+	}
+
+	sof_dai->configured = true;
+
+	return 0;
+}
+
+int hda_ctrl_dai_widget_free(struct snd_soc_dapm_widget *w)
+{
+	struct snd_sof_widget *swidget = w->dobj.private;
+	struct snd_soc_component *component = swidget->scomp;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct sof_ipc_dai_config *config;
+	struct snd_sof_dai *sof_dai;
+	struct sof_ipc_reply reply;
+	int ret;
+
+	sof_dai = swidget->private;
 
-		/* send DAI_CONFIG IPC */
-		return sof_ipc_tx_message(sdev->ipc, config->hdr.cmd, config, config->hdr.size,
-					  &reply, sizeof(reply));
+	if (!sof_dai || !sof_dai->dai_config) {
+		dev_err(sdev->dev, "No config for DAI %s\n", w->name);
```

Comment:
this error message is now identical to the one in `hda_ctrl_dai_widget_setup()`. Also I know that in some cases / files we prefix error messages with "error" in others we don't. Which one is this? :-) This message has no "error," but the one below has one... And in fact there's at least one more "no config for DAI" error case in `sdw_dai_config_ipc()` below and that one has an "error" word in it. So, it could be a separate patch / PR to improve that.